src/yggdrasil/switch.go

```diff
@@ -510,87 +529,106 @@ func (t *switchTable) _handleMsg(msg *switchMsg, fromPort switchPort, reprocessi
-		newTable.elems[pinfo.port] = tableElem{
+		elem := tableElem{
 			locator: loc,
 			port:    pinfo.port,
 			time:    pinfo.time,
 		}
+		newTable._insert(&elem)
+		newTable.elems[pinfo.port] = elem
 	}
-	newTable._msg = *t._getMsg()
 	t.core.peers.updateTables(t, &newTable)
 	t.core.router.updateTable(t, &newTable)
 }
 
-// Starts the switch worker
-func (t *switchTable) start() error {
-	t.core.log.Infoln("Starting switch")
-	// There's actually nothing to do to start it...
-	return nil
+func (t *lookupTable) _init() {
+	// WARNING: this relies on the convention that the self port is 0
+	self := tableElem{locator: t.self} // create self elem
+	t._start = self                    // initialize _start to self
+	t._insert(&self)                   // insert self into table
 }
 
-// Find the best port to forward to for a given set of coords
-func (t *lookupTable) lookup(coords []byte) switchPort {
-	var bestPort switchPort
-	myDist := t.self.dist(coords)
-	bestDist := myDist
-	var bestElem tableElem
-	for _, info := range t.elems {
-		dist := info.locator.dist(coords)
-		if dist >= myDist {
-			continue
-		}
+func (t *lookupTable) _insert(elem *tableElem) {
+	// This is a helper that should only be run during _updateTable
+	here := &t._start
+	for idx := 0; idx <= len(elem.locator.coords); idx++ {
+		refLoc := here.locator
+		refLoc.coords = refLoc.coords[:idx] // Note that this is length idx (starts at length 0)
+		oldDist := refLoc.ldist(&here.locator)
+		newDist := refLoc.ldist(&elem.locator)
 		var update bool
 		switch {
-		case dist < bestDist:
-			// Closer to destination
+		case newDist < oldDist: // new elem is closer to this point in the tree
 			update = true
-		case dist > bestDist:
-			// Further from destination
-		case info.locator.tstamp > bestElem.locator.tstamp:
-			// Newer root update
+		case newDist > oldDist: // new elem is too far
+		case elem.locator.tstamp > refLoc.tstamp: // new elem has a closer timestamp
 			update = true
-		case info.locator.tstamp < bestElem.locator.tstamp:
-			// Older root update
-		case info.time.Before(bestElem.time):
-			// Received root update via this peer sooner
+		case elem.locator.tstamp < refLoc.tstamp: // new elem's timestamp is too old
+		case elem.time.Before(here.time): // same dist+timestamp, but new elem delivered it faster
 			update = true
-		default:
 		}
 		if update {
-			bestPort = info.port
-			bestDist = dist
-			bestElem = info
+			here.port = elem.port
+			here.locator = elem.locator
+			here.time = elem.time
+			// Problem: here is a value, so this doesn't actually update anything...
+		}
+		if idx < len(elem.locator.coords) {
+			if here.next == nil {
+				here.next = make(map[switchPort]*tableElem)
+			}
+			var next *tableElem
+			var ok bool
+			if next, ok = here.next[elem.locator.coords[idx]]; !ok {
+				nextVal := *elem
+				next = &nextVal
+				here.next[next.locator.coords[idx]] = next
+			}
+			here = next
```

Comment:
This isn’t run for the last iteration of the enclosing loop (`<` vs `<=`)?

Reply:
Correct, this is setup for the next iteration of the loop. The last time through the loop, there's nothing to set up (`idx` would be out of range for `elem.locator.coords`), so this part is skipped.